New function to sample from frames (Issue #2419 )

nickeubank opened the pull request:
closes #2419 
Creates a function `.sample()` in generic.py to sample from pandas objects. Returns either a fixed number of rows or a share. Also includes a number of tests in `test_sample()` -- I am open to added suggestions. 

Input from users with panel experience appreciated!

nickeubank committed:
Add sample function with tests and docs

nickeubank commented:
Please excuse the many (meaningless) commits -- was learning to use git from commandline.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+    
+        Parameters
+        ----------
+            n: Number of rows to return. Cannot be used with frac.
```
please take another look at the numpy docstring standard:
https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt

In particular, you are indenting too much here and mixing up the lines for types/descriptions.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
         return self.iloc[-n:]
 
+  
+    def sample(self, n = None, frac = None , replace = False, weights = None, seed = None):
```
The PEP8 convention is to not use spaces around `=` when used in function signatures

hayd replied:
You might like to try [pep8radius](https://github.com/hayd/pep8radius) to fix all these.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+                    Default 'None' results in equal probability weighting.
+            seed: seed to be fed to numpy random.RandomState() Function. Default None. 
+        -----------
+        Note: sample sizes are passed to numpy.random.choice(), which accepts negative values. 
```
I don't think this is particularly desirable behavior to copy. It's confusing and not necessary. Instead, I would raise an error in such cases.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+                pass
+    
+            # If this is a dataframe and we get a string, check for column. 
+            elif type(self) == pd.core.frame.DataFrame and (type(weights) == str or type(weights) == unicode):
```
We prefer to avoid explicit types checks.

I would just use `if self.ndim > 1 and isinstance(weights, string_types)`

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+            elif type(self) == pd.core.frame.DataFrame and (type(weights) == str or type(weights) == unicode):
+                try: 
+                    weights = self[weights].values
+                except: 
```
Always catch explicit errors, if possible. Here you want to catch `KeyError`

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+        if weights != None:
+                                      
+            # If series, return ndarray
+            if type(weights) == pd.core.series.Series:
```
Instead of these convoluted checks, just use `np.asarray(weights)`, which converst series, lists and all other types convertable to arrays

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+        length = len(self)
+    
+        # Check weights for compliance
+        if weights != None:
```
It's more idiomatic to use `is not None` when comparing to `None`.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+        elif n == None and frac != None:
+            n = int(round(frac * length)) 
+        elif n != None and frac != None:
+            raise NotImplementedError('Please enter a value for "frac" OR "n", not both')
```
This should be `ValueError`. `NotImplementedError` should be reserved for code that is actually incomplete.

shoyer commented on pandas/tests/test_generic.py:
```diff
@@ -354,6 +354,52 @@ def test_head_tail(self):
+            setattr(o,axis,index(len(getattr(o,axis))))
+
+            # Check seeding -- run 10 times. 
+            for test in range(10):
```
What are you trying to do here? We don't need (or want) to verify numpy's random number generator. Instead, a few sanity checks would be fine.

nickeubank replied:
OK -- Looping through index types to ensure if something changed in .take() we wouldn't get problems. Stability through seeding seemed like a good way to make sure everything was being passed well. I don't really test randomness of np.choice -- just stability of results with a seed. Last one was to make sure weights passing correctly. That overkill I take it?

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+               Default = 5 if frac = None. 
+            frac: share of rows to return. Cannot be used with n. 
+            replace {True, False}: Sample with or without replacement.
+            weights: Accepts a Series, list, or ndarray of weights. If called on a 
```
Can weights be unnormalized?

shoyer commented on pandas/tests/test_generic.py:
```diff
@@ -354,6 +354,52 @@ def test_head_tail(self):
+    def test_sample(self):
+        o = self._construct(shape=10)
+        
+        # Run for all index types. 
```
Does any code here actually depend on the index type?

If not, I would skip these tests here. The `take` method should already be tested directly for all these different index types.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+    
+        # Setup vars
+        rs = np.random.RandomState(seed)
+        length = len(self)
```
I don't think much is gained by making this a separate variable -- I would probably still use `len(self)` throughout (just my two-cents).

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,84 @@ def tail(self, n=5):
+                    raise ValueError("Weights string not valid column name")
+                
+            # If any other type cases arise, throw error. 
+            elif type(weights) != np.ndarray:
```
Let numpy handle this itself by using `np.asarray` (as noted above)

shoyer commented:
This will also need documentation updates:

- An addition to "what's new" for v0.16.1 (not sure if this file has been created yet; best to wait until 0.16 is released)
- New documentation in an appropriate section of the docs
- Addition to the API docs (see `api.rst`).

nickeubank commented:
Thanks @shoyer. Adding changes and and testing.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,80 @@ def tail(self, n=5):
+
+    
+        # Check whether frac or N
+        if n == None and frac == None:
```
again, you want to be using `is None` here instead of `== None`

nickeubank replied:
ah right, sorry! Updating and testing before pushing.

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+            
+        Returns
+        -------
+        Frame of same type as caller. 
```
You can delete the "Frame of" here I think. Since it may not be a frame.

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+
+        Notes
+        -----
+        When called without arguments (.sample()), returns 5 random columns
```
rows, not columns. That would be a strange function :)

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+        elif n is None and frac is not None:
+            n = int(round(frac * len(self))) 
+        elif n is not None and frac is not None:
+            raise ValueError('Please enter a value for "frac" OR "n", not both')
```
I think variables are typically put in single backticks. [Here's](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#method-docstrings) all the rules.

nickeubank commented:
Doc additions soon -- wanted to pin down the function first, but may need to do some "real job" things before finishing.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+        
+        Parameters
+        ----------
+        n: in, optional, Default = 5 if frac = None. 
```
this should be `int` not `in`.

Also, the numpy docstring convention is usually *not* to list default values after the type. Rather, they should be described in the description part below.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+            Share of rows to return. Cannot be used with n. 
+        replace: boolean. Default = False. 
+            Sample with or without replacement. 
+        weights: Series, list, or ndarray. Default = None.
```
I would say `str or ndarray-like, optional`

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+            column as a string. Must be same length as index. 
+            Weights must be normalized. 
+        seed: int. Default = None. 
+            seed to be fed to numpy random.RandomState() Function.
```
I would just say `Seed for the random number generator.` The fact that we use numpy.RandomState under the covers is an implementation detail. Eventually, for example, we'll want this to work with groupby -- at which point we'll want to derive a separate seed for every group based on the seed the user provides.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+        When called without arguments (.sample()), returns 5 random rows
+        in a manner analogous to .head() or .tail(). 
+
+        -----------
```
this should go away

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+  
+    def sample(self, n=None, frac=None , replace=False, weights=None, seed=None):
+        """
+        Returns a sample of rows from object. 
```
The word "random" should appear in here somewhere.

Perhaps `Sample rows at random from this object`?

shoyer commented:
How about adding some tests to verify that the resulting frames/series have the right number of rows? These would also function as sanity checks to verify that sample is doing the right thing when you provide `frac` or `n` as arguments.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,76 @@ def tail(self, n=5):
+        n: int, optional.
+            Number of rows to return. Cannot be used with frac.
+            Default = 5 if frac = None. 
+        frac: float. 
```
sorry to be pedantic, but this should be `frac : float, optional` (one space before `:` and no period)

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,76 @@ def tail(self, n=5):
+            Default = 5 if frac = None. 
+        frac: float. 
+            Share of rows to return. Cannot be used with n. 
+            Default = None.
```
You don't need to explicitly give the default argument here -- it's already in the function signature.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,76 @@ def tail(self, n=5):
+                except(KeyError):
+                    raise KeyError("String passed to weights not a valid column name")
+            
+            if isinstance(weights, string_types) and self.ndim == 1:
```
rather than testing the type of `weights` twice, do these checks within a single if statement.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+                    except(KeyError):
+                        raise KeyError("String passed to weights not a valid column name")
+
+                elif self.ndim == 1:
```
this should just be an `else` here

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+        -------
+        Same type as caller. 
+
+        -----------
```
this line should be deleted.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+            If called on a DataFrame or Panel, will also accept the name of a 
+            column as a string. Must be same length as index. 
+            Weights must be normalized. 
+        seed : int
```
this should be `int, optional`

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+        n : int, optional
+            Number of rows to return. Cannot be used with frac.
+            Default = 5 if frac = None. 
+        frac : float 
```
frac and replace should both be marked as `optional`

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+        frac : float 
+            Share of rows to return. Cannot be used with n. 
+        replace : boolean
+            Sample with or without replacement. Default = False. 
```
I usually let function signatures document default values like `False` instead of repeating it in the docstring, but that's not a strong rule.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
         return self.iloc[-n:]
 
+  
+    def sample(self, n=None, frac=None , replace=False, weights=None, seed=None):
```
pep8: extra space after `frac=None`

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+            # Strings acceptable if not a series
+            if isinstance(weights, string_types): 
+
+                if self.ndim >1 :
```
pep8: should be `if self.ndim > 1:`

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,79 @@ def tail(self, n=5):
+                if self.ndim >1 :
+                    try:
+                        weights = self[weights]
+                    except(KeyError):
```
pep8: `except` is a keyword, not a function. This should be `except (KeyError)` or better yet `except KeyError`  (since you are only allowing one error).

shoyer commented:
I have a few style/doc nits but this looks pretty good to me, pending documentation updates.

nickeubank commented:
Great, thanks @shoyer. Once the 16.1 `what's new` is up I'll write up docs.

jreback commented:
whats new is now available for 0.16.1 https://github.com/pydata/pandas/commit/5ebf521eac528f5a4b772e4677f70b689b8b1826

nickeubank commented:
Hi All: I added a note to whatsnew, and added DataFrame.sample and Series.sample everywhere that seemed relevant in the api.rst docs. 

Is there anything more I should be adding / can anyone point me to more detailed docs on contributing to documentation besides the readme.rst in pandas/docs?

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,78 @@ def tail(self, n=5):
+        Returns
+        -------
+        Same type as caller. 
+        -----------
```
these final `-----------` should be removed

shoyer commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,78 @@ def tail(self, n=5):
+        if n < 0:
+            raise ValueError("A negative number of rows requested. Please provide positive value.")
+    
+        locs = rs.choice(len(self), size = n, replace = replace, p = weights)
```
Please double check PEP 8 -- you should not be using spaces around `=` in function arguments :). This should be `rs.choice(len(self), size=n, replace=replace, p=weights)`, and the same elsewhere.

shoyer commented:
@nickeubank It's really up to you where this belong in the documentation! I might add a new section "Random sampling" to the "Indexing and selecting" portion of the documentation:
http://pandas.pydata.org/pandas-docs/version/0.16.0/indexing.html

You can find the source for this section in `doc/source/indexing.rst`

nickeubank commented:
@shoyer -- oh! I thought that only applied to defining functions, I didn't realize that the no spaces applied in function calls. adding now.

shoyer commented on doc/source/whatsnew/v0.16.1.txt:
```diff
@@ -16,8 +16,41 @@ We recommend that all users upgrade to this version.
+.. _whatsnew_0161.enhancements.sample:
 
+Sample
+^^^^^^^^^^^^^^^^
```
The number of characters in this line should match the length of the preceding line, e.g., you only need `^^^^^^`

shoyer commented on doc/source/indexing.rst:
```diff
@@ -500,6 +500,62 @@ A list of indexers where any element is out of bounds will raise an
+   s.sample()
+   
+   # One may specify either a number of rows:
+   s.sample(n = 3)
```
You fixed the spacing around `=` in your code but not the documentation examples :)

shoyer commented:
@nickeubank sorry for the delay looking at this again -- github doesn't send any notifications when a PR is updated, only when new comments appears.

This looks very good to me -- pretty much good to go, once you fix a few styling issues.

@TomAugspurger @jorisvandenbossche @cpcloud @jreback Any additional comments?

jreback commented:
pls rebase

jreback commented on doc/source/indexing.rst:
```diff
@@ -500,6 +500,62 @@ A list of indexers where any element is out of bounds will raise an
+
+
+By default, each row has an equal probability of being selected, but if you want rows
+to have different probabilities, you can pass ``sample`` sampling weights. This weights
```
I think u mean to say weights here (as that is the argument name) (or maybe just need to put back ticks around weights)

jreback replied:
I would say (and change in the code) that weights will be normalized to sum to 1 ; I suppose a nan weight excludes that weight?

jreback replied:
eg you can do a fillna(0) on the weights array

jreback replied:
need to add a test for this type of nan handling

jorisvandenbossche commented on doc/source/indexing.rst:
```diff
@@ -500,6 +500,62 @@ A list of indexers where any element is out of bounds will raise an
+	
+.. ipython :: python
+
+   df = DataFrame({'col1':[9,8,7,6], 'weight_column':[0.5, 0.4, 0.1, 0]})
```
You are redefining a `df` here that is used later on (line 610), but that example there was based on how `df` was defined above (line 166). So can you give this here another name (eg `df2`), or repeat the definition of `df` from above below.

You can see the build log of the docs in the travis build (third job): https://travis-ci.org/pydata/pandas/jobs/56058495, scroll down and expand the `$ if [ -f /tmp/doc.log ]; then cat /tmp/doc.log; fi` line

jorisvandenbossche commented on doc/source/whatsnew/v0.16.1.txt:
```diff
@@ -16,8 +16,41 @@ We recommend that all users upgrade to this version.
+Sample
+^^^^^^^^^^^^^^^^
 
+Series, DataFrames, and Panels now have a new method: :meth:`~pandas.core.sample`.
```
the `pandas.core.sample` should be `pandas.Series.sample` (or with DataFrame), as this is how it is exposed publicly.

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -1940,6 +1940,77 @@ def tail(self, n=5):
+        Parameters
+        ----------
+        n : int, optional
+            Number of rows to return. Cannot be used with frac.
```
Can you do backticks around frac: like `` `frac` `` (the it is more clear it is a keyword argument, the same for `n` two lines below)

jorisvandenbossche commented:
This is becoming a really nice improvement!

Two questions:

- Now, `axis=0` is passed to the eventual `self.take(..)` call. Should this be configurable? (by having an `axis` argument in `sample`) E.g. for panels, I don't use them, but for those who do, maybe you want to be able to specify on which axis to sample?

- The default of `n=5`. This was discussed before, and there already some objections then. But still, this seems a bit strange to me. E.g. also the defaults of returning 1 sample (`n=1`, this is what `numpy.choice` does), or returning all rows in random order (`n=len(self)`) both do seem as sensible defaults to me.

sinhrks commented:
It is nice if ``seed`` can accept ``np.random.RandomState`` also (like scikit-learn's ``random_state`` does). And I prefer keyword ``random_state`` to ``seed`` if there is no specific reason.

jorisvandenbossche commented:
Any other comments on the `axis`, default of `n` and `random_state` questions?

jreback commented:
- ``axis`` should prob added (and default to 0).
- ``seed`` should accept a random_state object / int (is this passable to ``np.random.random_state``?)

shoyer commented:
scikit learn has a really nice little utility function, `check_random_state`: https://github.com/scikit-learn/scikit-learn/blob/2b3fabc/sklearn/utils/validation.py#L498

If we want a comprehensive solution, we should just directly copy it into pandas and use it to process the `seed` argument (or, perhaps we should also call it `random_state` [like scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.RandomizedPCA.html#sklearn.decomposition.RandomizedPCA)). scikit-learn is BSD licensed, some we'll just need to add a copy of its license to the `LICENSES` directory.

nickeubank commented:
Thanks all -- traveling for conferences this week and next but these all look like great ideas, and i'll integrate them when I'm back!

jreback commented:
@nickeubank can you rebase and squash this.

nickeubank commented:
Thanks all for feedback. Now rebashed and squashed, with the following changes (and associated tests):
 - added axis argument (default 0). 
 - seed is now called random_state and accepts either integers or numpy.random.RandomState objects.
 - weights will now re-normalize weights that do not sum to one, and will set values of np.nan or None to zero. 
 - default is now to return 1 row (as per wisdom of the crowd). 
 - documentation updated.

shoyer commented:
@nickeubank looks like you haven't pushed those fixes yet?

nickeubank commented:
Sorry, noticed typo in documentation -- recompiling, then will push.

nickeubank commented:
OK, now pushed. Sorry for delay.

shoyer commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+        # random_state may be either a RandomState object, 
+        # integer, or None. Otherwise raise an error. 
+     
+        if isinstance(random_state, int):
```
Can you factor this out into a separate helper function? Could put it in `pandas.core.common`

shoyer commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+            # No infs allowed. The np.nan_to_num() command below would make these large values
+            # which is pretty unintuitive. 
+            if np.isinf(weights).any():
+                raise ValueError("weight vector many not include `inf` values")
```
spelling: many -> may

shoyer commented on setup.py:
```diff
@@ -197,7 +197,7 @@ def build_extensions(self):
 ISRELEASED = False
 VERSION = '%d.%d.%d' % (MAJOR, MINOR, MICRO)
-QUALIFIER = ''
+QUALIFIER = 'rc1'
```
you don't really want to change this :)

shoyer commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -14,6 +14,7 @@ Highlights include:
 - Backwards incompatible change to ``Timedelta`` to conform the ``.seconds`` attribute with ``datetime.timedelta``, see :ref:`here <whatsnew_0160.api_breaking.timedelta>`
 - Changes to the ``.loc`` slicing API to conform with the behavior of ``.ix`` see :ref:`here <whatsnew_0160.api_breaking.indexing>`
 - Changes to the default for ordering in the ``Categorical`` constructor, see :ref:`here <whatsnew_0160.api_breaking.categorical>`
+<<<<<<< HEAD
```
you don't want the changes in this file at all

shoyer commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+        # Processing random_state argument
+        ###
+    
+        # random_state may be either a RandomState object, 
```
I think you're going a little heavy on the comments here (and throughout) -- generally speaking, comments should only exist to clarify things that are not obvious from reading the code.

jreback commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+        
+        # This checks validity of axis argument.
+        if axis > (self.ndim - 1):
+            raise ValueError("Axis argument out of range")
```
use ``axis = self._get_axis_number(axis)`` will raise (and translate such things as ``axis='columns'`` to the correct axis number

jreback commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+            Seed for the random number generator (if int), or numpy RandomState 
+            object.
+        axis : int, optional
+            Axis to sample. Default = 0.
```
indicate that will accept index/columns as well

jreback commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+                    raise ValueError("Strings cannot be passed as weights when sampling from a Series.")
+
+            #normalize format of weights to ndarray. 
+            weights = np.asarray(weights, dtype = np.float)
```
use ``np.float64``

jreback replied:
make this a Series.

nickeubank replied:
I have to pass this to numpy.choice later, which doesn't want a series, so this seemed more straightforward.

jreback replied:
who say it doesn't accept a Series.
Try to use pandas functions as much as possible. Eg. then you don't have to use odd things like ``np.nan_to_num`` (which doesn't work for most dtypes)
```
In [88]: np.random.choice(Series([1,2,3]))
Out[88]: 2
```

nickeubank replied:
ah, ok. Will update.

jreback commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+                raise ValueError("weight vector many not include negative values")
+
+            # If has nan, set to zero. Already know there are no infs. 
+            weights = np.nan_to_num(weights)            
```
``weights = weights.fillna(0)``

nickeubank replied:
As above, just wanted to avoid moving into a Series, then back to an array for np.choice().

jreback replied:
do you have a valid reason for this?

nickeubank replied:
Just that it seemed smoother to only make one type conversion into a numpy array then move into a Series, clean, then put into numpy array for np.choice().

jreback commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+
+
+            # Check that weights sum to 1. If not, renormalize. 
+            if np.sum(weights) != 1:
```
``weights.sum()``

jreback commented on pandas/tests/test_generic.py:
```diff
@@ -354,6 +354,155 @@ def test_head_tail(self):
+    def test_sample(self):
+        o = self._construct(shape=10)
+
+        ###
```
add the issue number as a comment here

nickeubank commented:
OK, think i've made all requested changes. Let me know if it looks ok and whether I should rebase / squash again!

nickeubank commented:
There a way to re-start a build without pushing a new commit? Seems to be an ip failure, not real problem...

nickeubank commented:
@jreback @shoyer Sorry to bug you guys, but do you happen to know what's going on with this last build? I think the problem is that o.iloc[0:1] is returning a Series rather than the original data type of o, but only during some builds (and not on my own computer). Is there an easy way to pull the first row of an object that preserves type for this type of comparison?

jorisvandenbossche commented on doc/source/indexing.rst:
```diff
@@ -508,6 +508,81 @@ A list of indexers where any element is out of bounds will raise an
+   s.sample()
+   
+   # One may specify either a number of rows:
+   s.sample(n = 3)
```
Can you keep to PEP 8 here? (so `n=3` without spaces)

jorisvandenbossche replied:
and below as well

jorisvandenbossche commented on doc/source/indexing.rst:
```diff
@@ -508,6 +508,81 @@ A list of indexers where any element is out of bounds will raise an
+	
+	# Weights will be re-normalized automatically
+	example_weights2 = [0.5, 0, 0, 0, 0, 0]
+	s.sample(n=1, weights= example_weights2)
```
your indentation here should be the same as the first block

jorisvandenbossche commented:
@nickeubank I am adding a couple more of stylistic comments. But, I think you are mixing tabs and spaces for indentation. Can you check your editor to only use spaces? At least in the docs page there are some tabs

jorisvandenbossche commented on doc/source/whatsnew/v0.16.1.txt:
```diff
@@ -20,6 +20,7 @@ Highlights include:
 
 Enhancements
 ~~~~~~~~~~~~
+.. _whatsnew_0161.enhancements.sample:
```
this one should be at the line before the 'sample' title

jorisvandenbossche commented on doc/source/whatsnew/v0.16.1.txt:
```diff
@@ -135,6 +136,45 @@ values NOT in the categories, similarly to how you can reindex ANY pandas index.
+Sample
+^^^^^^^^^^^^^^^^
+
+Series, DataFrames, and Panels now have a new method: :meth:`~pandas.core.sample`.
```
``~pandas.core.sample`` -> ``~pandas.DataFrame.sample``

jorisvandenbossche commented on doc/source/whatsnew/v0.16.1.txt:
```diff
@@ -135,6 +136,45 @@ values NOT in the categories, similarly to how you can reindex ANY pandas index.
+
+   example_series = Series([0,1,2,3,4,5])
+
+   # When no arguments are passed, returns 5 rows like .head() or .tail()
```
this is now not correct anymore

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,119 @@ def tail(self, n=5):
+            Number of rows to return. Cannot be used with `frac`.
+            Default = 1 if `frac` = None. 
+        frac : float, optional
+            Share of rows to return. Cannot be used with `n`. 
```
'share' -> 'fraction' ? (that's how it is called in the other docs)

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,119 @@ def tail(self, n=5):
+
+                if self.ndim > 1 :
+                    try:
+                        weights = self[weights]
```
this should access the 'weights' column in the correct axis. Eg if you sample a dataframe it's columns (`axis=1`), should it then be possible to give a row index name? 
In any case, now this can trigger the "Weights and axis to be sampled must be of same length" error when the dataframe is not square (and would give faulty results if it would be square)

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,119 @@ def tail(self, n=5):
+                raise ValueError("Weights and axis to be sampled must be of same length")
+
+            # No infs allowed. The np.nan_to_num() command below would make these large values
+            # which is pretty unintuitive. 
```
I think you don't use `nan_to_num` anymore?

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,119 @@ def tail(self, n=5):
+                    raise ValueError("Strings cannot be passed as weights when sampling from a Series.")
+
+            #normalize format of weights to ndarray. 
+            weights = pd.Series(weights, dtype = 'float64')
```
I think @jreback said to make this a Series, but if you take the `.values` of it (line 2040 below), I don't really see the point of that?

jorisvandenbossche commented:
Some other comments / questions:

* When passing a series as `weights`, should the values of the series be aligned on the index? (this is not the case at the moment)

* Should the default `n=1` case for DataFrames return a 1-row DataFrame, or that row as a Series (and for sampling a Series -> return 1-element Series or a scalar)? As a comparison `np.random.choice` reduces the dimensionality for `n=1`, but in the case of pandas you then loose the info of the index. So maybe keep it as is (DataFrame -> 1-row DataFrame)?

* I know nothing about Panels (never used it), but should the default axis be 0? (for DataFrame's the info axis is 1, so by default samples from axis 0. But for panels the info axis is 0). Are there similar functions to look at to do this consistently?

* The `weights` don't work for Panels at the moment (and I also don't see how it could work?). But if this is by design, it should get an appropriate error message, testing and docs (the docstring now says: "If called on a DataFrame or Panel, will also accept the name of a column as a string" but this does not work)

nickeubank commented:
Thanks @jorisvandenbossche!

- "should one row of df be returned as df or series?" I really dislike the idea that different arguments (`n=1` versus `n=2`) return different types of objects, so I strongly prefer it returning a one row DataFrame if that's the item being sampled. 

- Seems like to get the string-of-weights to work, I need to be able to index by label from a given axis. Is there a way to do this? `.loc[]` doesn't seem to have an "axis" argument. I could set it up with some ugly code to do `.loc[:,:,'weight_string']` with `'weight_string'` in the position associated with the relevant axis, but doesn't feel pythonic. Is there another way? 

- I also know nothing about panels -- I was assuming that axis 0 and 1 are same as DataFrame and `axis=2` is just time, but sounds like that's incorrect? Input by someone who uses panels would be appreciated. 

- the `weights = weights.values` line does negate much of the value of using Series, but I've found that `np.choice()` is not consistent in its ability to work with Series as weights. When I tried passing a Series as weights, I found that it worked on my computer, but 3 of the 5 Travis builds kept failing. The `weights = weights.values` is the only way I could fix the problem. 
 - should the values of series be aligned: given this issue, I'm not sure how I'd do that since `np.choice` won't take a Series (see discussion below). If you can suggest a way to do this, I'd be happy to do so. 
 
Small things fixed:
- tabs were causing issues with format. All spaces now. 
- PEP-8 in function signatures fixed. Sorry about that.

nickeubank commented:
Thanks @jorvisvandenbossche! 

- "should one row of df be returned as df or series?" I really dislike the idea that different arguments (`n=1` versus `n=2`) return different types of objects, so I strongly prefer it returning a one row `DataFrame` if that's the item being sampled. 

- I think it makes sense to just allow for strings to be passed as weights for `DataFrames` when sampling from rows. I starting writing it in a more general form to allow for rows as weights, but it gets very hairy fast, especially because rows may be of mixed types. Since sampling from rows is likely the most common use case, and one can just explicitly pass a row as a weight if needed, I think this is sufficient. Changing documentation accordingly, and tests added for appropriate errors added for panels.  

- I also know nothing about `panels` -- I was assuming that axis 0 and 1 are same as `DataFrame` and `axis=2` is just time, but sounds like that's incorrect? Input by someone who uses panels would be appreciated. 

- the `weights = weights.values` line does negate much of the value of using Series, but I've found that `np.choice()` is not consistent in its ability to work with `Series` as weights. When I tried passing a Series as weights, I found that it worked on my computer, but 3 of the 5 Travis builds kept failing. The `weights = weights.values` is the only way I could fix the problem. 
 - should the values of series be aligned: given `np.choice` won't take a Series, how would I go about this? If you can suggest a way to do this, I'd be happy to do so. 
 
Small things fixed:
- tabs were causing issues with format. All spaces now. 
- PEP-8 in function signatures fixed. Sorry about that.

sinhrks commented on doc/source/api.rst:
```diff
@@ -713,6 +714,7 @@ Indexing, iteration
    DataFrame.where
    DataFrame.mask
    DataFrame.query
+   DataFrame.sample
```
I understand the duplication is intentional, but only adding to "... Selection ... " section may be better? I understand the function doesn't do indexing or iteration.

sinhrks commented on doc/source/whatsnew/v0.16.1.txt:
```diff
@@ -135,6 +135,47 @@ values NOT in the categories, similarly to how you can reindex ANY pandas index.
 
 See the :ref:`documentation <advanced.categoricalindex>` for more. (:issue:`7629`)
 
+.. _whatsnew_0161.enhancements.sample:
```
Can you add highlights description, linking to this section, like:

- https://github.com/pydata/pandas/blob/master/doc/source/whatsnew/v0.16.1.txt#L12

sinhrks commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,120 @@ def tail(self, n=5):
+  
+    def sample(self, n=None, frac=None, replace=False, weights=None, random_state=None, axis = 0):
+        """
+        Returns a random sample of rows from object. 
```
Along axis?

jorisvandenbossche commented on doc/source/indexing.rst:
```diff
@@ -508,6 +508,81 @@ A list of indexers where any element is out of bounds will raise an
+------------------------
+.. versionadded::0.16.1
+
+A random selection of rows or columns from a Series, DataFrame, or Panel with the ``.sample()`` method. The method will sample rows by default, and accepts a specific number of rows/columns to return, or a fraction of rows. 
```
can you replace ``` ``.sample()`` ```` with `` :meth:`~DataFrame.sample` `` (of `` :meth:`Series.sample` ``), the result will be almost the same, but also be a link to the api docstring page

jorisvandenbossche commented:
@jreback could you shed some light on the panel issues?

@nickeubank 

- agree on the `n=1` and `n>1` giving the same output type
- OK for me on limiting specifying `weights` as a string for now to DataFrame columns
- on the `weights = weights.values` thing, it just seems to me that if you want to end up with an array, the initial `weights = pd.Series(weights, dtype = 'float64')` is not needed?

jreback commented on doc/source/whatsnew/v0.16.1.txt:
```diff
@@ -135,6 +136,47 @@ values NOT in the categories, similarly to how you can reindex ANY pandas index.
+
+Sample
+^^^^^^^^^^^^^^^^
+
```
needs to be exactly length of the text

jreback commented on pandas/core/common.py:
```diff
@@ -3319,3 +3319,31 @@ def _maybe_match_name(a, b):
+    np.random.RandomState
+    """
+
+    if isinstance(state, int):
```
use ``is_integer(...)`` (same result, just internally more consistent)

jreback commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,120 @@ def tail(self, n=5):
+        axis = self._get_axis_number(axis)
+    
+        # Store length of relevant axis of object. 
+        axis_length = self.shape[axis]    
```
you can eliminate some of these comments. Too much verbiage here.

jreback commented:
some minor code comments.

if an axis is not provided, it is usual to use ``axis = self._stat_axis``, this will yield:
``Series->0, DataFrame->0, Panel->1``. This type of code is used all over ``generic.py``

pls squash.
otherwise lgtm. once you have made changes, ping when green.

nickeubank commented:
@jreback : I think we're good to go. 

Added axis = self._stat_axis and associated tests; added is_integer() to _random_state(); thinned comments; fixed format in "what's new"; updated docs on only supporting column names for dataframes. 

@jreback @shoyer  @jorisvandenbossche @TomAugspurger @sinhrks : thank you all for your input on this -- it's been a great learning experience, and hopefully a useful new feature!

shoyer commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,120 @@ def tail(self, n=5):
+                    raise ValueError("Strings cannot be passed as weights when sampling from a Series or Panel.")
+
+            #normalize format of weights to Series. 
+            weights = pd.Series(weights, dtype = 'float64')
```
PEP8 would be `weights = pd.Series(weights, dtype='float64')`

shoyer commented on pandas/core/generic.py:
```diff
@@ -1948,6 +1948,120 @@ def tail(self, n=5):
+        """
+        
+        ###        
+        # Process axis argument
```
I still think this function would be clearer if you removed the all comments. Yes, really! All your comments are a pretty literal restatements of what the code itself does.

jorisvandenbossche replied:
@nickeubank maybe you can already remove the `###` and blank line after it to make it visually a bit less heavy.

jorisvandenbossche commented:
@nickeubank Looks good to me! 
Certainly a very usefull feature. We have been very 'commenting' here (but I think that is good, to deliver good quality, which it is!)

sinhrks commented:
Thanks for the change, great job :)

jreback commented:
merged via https://github.com/pydata/pandas/commit/8f0f417cbe67e6a704d24993aa76030fa84bb595

@nickeubank thanks for effort! and responding to comments. all of this just makes pandas better!

shoyer commented:
woohoo, thanks!
